app/services/events/post_process_service.rb

```diff
@@ -0,0 +1,150 @@
+
+      @customer = if event.external_subscription_id
+        organization.subscriptions.find_by(external_id: event.external_subscription_id)&.customer
+      else
+        Customer.find_by(external_id: event.external_customer_id, organization_id: organization.id)
+      end
+    end
+
+    def subscriptions
+      return @subscriptions if defined? @subscriptions
+
+      subscriptions = if customer && event.external_subscription_id.blank?
+        customer.subscriptions
+      else
+        organization.subscriptions.where(external_id: event.external_subscription_id)
+      end
+      return unless subscriptions
+
+      @subscriptions = subscriptions
+        .where("date_trunc('second', started_at::timestamp) <= ?::timestamp", event.timestamp)
+        .where(
+          "terminated_at IS NULL OR date_trunc('second', terminated_at::timestamp) >= ?",
+          event.timestamp,
+        )
+        .order('terminated_at DESC NULLS FIRST, started_at DESC')
+    end
+
+    def billable_metric
+      @billable_metric ||= organization.billable_metrics.find_by(code: event.code)
+    end
+
+    def expire_cached_charges(subscriptions)
+      active_subscription = subscriptions.select(&:active?)
+      return if active_subscription.blank?
+      return unless billable_metric
+
+      charges = billable_metric.charges
+        .joins(:plan)
+        .where(plans: { id: active_subscription.map(&:plan_id) })
+
+      charges.each do |charge|
+        active_subscription.each do |subscription|
+          Subscriptions::ChargeCacheService.new(subscription:, charge:).expire_cache
+        end
+      end
+    end
+
+    def quantified_event_service
+      @quantified_event_service ||= QuantifiedEvents::CreateOrUpdateService.new(event)
+    end
+
+    def should_handle_quantified_event?
+      quantified_event_service.matching_billable_metric?
+    end
+
+    def handle_quantified_event
+      service_result = quantified_event_service.call
+      service_result.raise_if_error!
+
+      # TODO: Remove this relation
+      event.quantified_event_id = service_result.quantified_event&.id
+      event.save!
+    end
+
+    def handle_pay_in_advance
+      return unless billable_metric
+
+      charges.where(invoiceable: false).each do |charge|
+        Fees::CreatePayInAdvanceJob.perform_later(charge:, event:)
+      end
+
+      # NOTE: ensure event is processable
+      return if !billable_metric.count_agg? && event.properties[billable_metric.field_name].nil?
+
+      charges.where(invoiceable: true).each do |charge|
+        Invoices::CreatePayInAdvanceChargeJob.perform_later(charge:, event:, timestamp: event.timestamp)
+      end
+    end
+
+    def charges
+      return Charge.none unless subscriptions.first
+
+      subscriptions
+        .first
+        .plan
+        .charges
+        .pay_in_advance
+        .joins(:billable_metric)
+        .where(billable_metric: { code: event.code })
+    end
+
+    def billable_metric
```

Comment:
Lint/DuplicateMethods: Method `Events::PostProcessService#billable_metric` is defined at both app/services/events/post_process_service.rb:72 and app/services/events/post_process_service.rb:136.